test/spec/tokenManager.js

```diff
@@ -1045,6 +1046,102 @@ describe('TokenManager', function() {
       jest.advanceTimersByTime(1000);
       expect(callback).toHaveBeenCalledWith('test-idToken', tokens.standardIdTokenParsed);
     });
+
+    describe('too many renew requests', () => {
+      it('should emit too many renew error when latest 10 expired event happen in 30 seconds', () => {
+        setupSync({
+          tokenManager: { autoRenew: true }
+        });
+        client.tokenManager.renew = jest.fn().mockImplementation(() => Promise.resolve());
+        const handler = jest.fn().mockImplementation(err => {
+          util.expectErrorToEqual(err, {
+            name: 'AuthSdkError',
+            message: 'Too many token renew requests',
+            errorCode: 'INTERNAL',
+            errorSummary: 'Too many token renew requests',
+            errorLink: 'INTERNAL',
+            errorId: 'INTERNAL',
+            errorCauses: []
+          });
+        });
+        client.tokenManager.on('error', handler);
+        let startTime = Math.round(Date.now() / 1000);
+        // 2 * 10 < 30 => emit error
+        for (let i = 0; i < 10; i++) {
+          util.warpToUnixTime(startTime);
+          client.emitter.emit('expired');
+          startTime = startTime + 2;
+        }
+        expect(handler).toHaveBeenCalledTimes(1);
+        expect(client.tokenManager.renew).toHaveBeenCalledTimes(9);
+      });
+
+      it('should keep emitting errors if expired events keep emitting in 30s', () => {
+        setupSync({
+          tokenManager: { autoRenew: true }
+        });
+        client.tokenManager.renew = jest.fn().mockImplementation(() => Promise.resolve());
+        const handler = jest.fn();
+        client.tokenManager.on('error', handler);
+        let startTime = Math.round(Date.now() / 1000);
+        // 2 * 10 < 30 => emit error
+        for (let i = 0; i < 20; i++) {
+          util.warpToUnixTime(startTime);
+          client.emitter.emit('expired');
+          startTime = startTime + 2;
+        }
+        expect(handler).toHaveBeenCalledTimes(11);
+        expect(client.tokenManager.renew).toHaveBeenCalledTimes(9);
+      });
+  
+      it('should not emit error if time diff for the latest 10 requests are more than 30s', () => {
+        setupSync({
+          tokenManager: { autoRenew: true }
+        });
+        const handler = jest.fn();
+        client.tokenManager.on('error', handler);
+        client.tokenManager.renew = jest.fn().mockImplementation(() => Promise.resolve());
+        let startTime = Math.round(Date.now() / 1000);
+        // 5 * 10 > 30 => not emit error
+        for (let i = 0; i < 20; i++) {
+          util.warpToUnixTime(startTime);
+          client.emitter.emit('expired');
+          startTime = startTime + 5;
+        }
```

Comment:
Why does this say 5*10?  Isn't this 5*19?

the rest below is more general commentary.  I had a hard time figuring out what this is trying to do because of the vague variable names and logic without indication of what it was logic-ing.
```
let startTime = Math.round(Date.now() / 1000); 
for(let notExpiredYet = startTime; notExpiredYet < (startTime+30)); notExpiredYet += 5) {
    util.warpToUnixTime(notExpiredYet);
    client.emitter.emit('expired'); // should not trigger effects
}
```

Or something like this - use the for loop and variable names to explain what you're doing.

Another alternative:
```
let startTime = Math.round(Date.now() / 1000);
for( let offset of [0,5,10,15,20,25] ) { 
   util.WarpToUnixTime(startTime+offset);
   client.emitter.emit(expired); // should not trigger effects
}
```
or
```
let startTime = Math.round(Date.now() / 1000);
let offset = 0;
while( offset < 30 ) { 
   util.WarpToUnixTime(startTime+offset);
   client.emitter.emit(expired); // should not trigger effects
   offset += 5;
}
```